server/src/main/java/org/opensearch/cluster/coordination/JoinTaskExecutor.java

```diff
@@ -466,13 +476,33 @@ public static void ensureMajorVersionBarrier(Version joiningNodeVersion, Version
         }
     }
 
+    public static void ensureNodeCommissioned(DiscoveryNode node, Metadata metadata) {
+        DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.custom(DecommissionAttributeMetadata.TYPE);
+        if (decommissionAttributeMetadata != null) {
+            DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+            DecommissionStatus status = decommissionAttributeMetadata.status();
+            if (decommissionAttribute != null && status != null) {
+                // We will let the node join the cluster if the current status is not IN_PROGRESS or SUCCESSFUL
+                if (node.getAttributes().get(decommissionAttribute.attributeName()).equals(decommissionAttribute.attributeValue())
+                    && (status.equals(DecommissionStatus.IN_PROGRESS) || status.equals(DecommissionStatus.SUCCESSFUL))) {
```

Comment:
For my understanding what happens on FAILURE, do we want decommission statue to be reflected in cluster health API as well? I feel it would be good if we can add that there.

Reply:
> For my understanding what happens on FAILURE

We are marking the status as `FAILED`, whenever any step after `INIT` fails. For eg., when after node removal executor execute state update, and observer is waiting for removal of those nodes but still couldn't get the expected state after timeout. Another example is removing to-be-decommissioned leader from Voting config, etc.

> do we want decommission statue to be reflected in cluster health API as well? I feel it would be good if we can add that there.

Opened a new issue to discuss this change - https://github.com/opensearch-project/OpenSearch/issues/4528